core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala

```diff
@@ -889,21 +892,42 @@ private class PythonAccumulatorParam(@transient private val serverHost: String,
     socket
   }
 
-  override def zero(value: JList[Array[Byte]]): JList[Array[Byte]] = new JArrayList
+  override def reset(): Unit = {
+    this._acc = Collections.synchronizedList(new JArrayList[Array[Byte]])
+  }
+
+  override def isZero: Boolean = {
+    this._acc.isEmpty
+  }
+
+  override def copyAndReset(): PythonAccumulatorV2 = new PythonAccumulatorV2(serverHost, serverPort)
+
+  override def copy(): PythonAccumulatorV2 = {
+    val newAcc = new PythonAccumulatorV2(serverHost, serverPort)
+    newAcc._acc.addAll(this._acc)
+    newAcc
+  }
+
+  // This happens on the worker node, where we just want to remember all the updates
+  override def add(val2: JList[Array[Byte]]): Unit = {
+    _acc.addAll(val2)
+  }
+
 
-  override def addInPlace(val1: JList[Array[Byte]], val2: JList[Array[Byte]])
-      : JList[Array[Byte]] = synchronized {
+  override def merge(other: AccumulatorV2[JList[Array[Byte]], JList[Array[Byte]]]): Unit = {
+    val otherPythonAccumulator = other.asInstanceOf[PythonAccumulatorV2]
+    // This conditional isn't strictly speaking needed - merging only currently happens on the
+    // driver program - but that isn't gauranteed so incase this changes.
     if (serverHost == null) {
-      // This happens on the worker node, where we just want to remember all the updates
-      val1.addAll(val2)
-      val1
+      // We are on the worker
+      add(otherPythonAccumulator._acc)
     } else {
       // This happens on the master, where we pass the updates to Python through a socket
       val socket = openSocket()
       val in = socket.getInputStream
       val out = new DataOutputStream(new BufferedOutputStream(socket.getOutputStream, bufferSize))
-      out.writeInt(val2.size)
-      for (array <- val2.asScala) {
+      out.writeInt(otherPythonAccumulator._acc.size)
```

Comment:
This is looking good. There is one more possible thread-safety issue here. Here we access the size of another (synchronized) list, then iterate over it. Both could be a problem if the list is changed somewhere during this process. I think we'd want to explicitly synchronize on `otherPythonAccumulator._acc` for this whole block, because that's what `Collections.synchronizedList` is doing in all other cases to guard access to it.

Either that or confirm that this would never be called in a multi-threaded context.